### ISSUE TITLE
Make kata work together with device plugin to pin vCPU threads to isolated cpu cores

### ISSUE THREAD

**Issue by fredliu-db** [kata-containers/kata-containers#4045]
**Is your feature request related to a problem? Please describe.**
Background:
Some workloads are both computation sensitive and IO sensitive. For such workloads, we need to reserve some cores on a host for IO (disk, network), reserve other cores for host processes and other side car pods and reserve all remaining cores for work load pods. Due to reducing context switching and isolating unnecessary interruption, cpu isolation and vcpu threads pinning could [significantly improve workload performance.](https://github.com/kata-containers/kata-containers/issues/3726)

Sample use case:
On a 64 cores host, cores 0 - 54 are reserved for kata workload pods, cores 55-62 are reserved for OS and other processes, core 63 and 64 are reserved for processing IO requests.  Kata pod 1 requests 2 cores, then it should be assigned to core 0-1, and pin all its threads to core 0-1. Kata pod 2 requests 4 cores, then it should be assigned to core 2-5, and pin all its threads to cores 2-5.

There are three requirements for assigning a kata pod to such host:
1) Kata shim needs to know the core numbers to pin threads.
2) Maintaining cpu usage information on host, so that different kata pods will be assigned to dedicated cores.
3) Maintaining resource allocation information, to make k8s be able to schedule pods to hosts. 

Hence device plugin comes into the picture.  We will discuss how device plugin communicates with kata and how kata responds to it.

**Describe the solution you'd like**
Create a [device plugin](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/), the device plugin will isolate some cpus on host based on workload requirements. CPUs reserved for workload pods will be exposed to k8s as devices. Device plugin has a metadata store to track which cpu cores have been assigned and uses the store to recycle cores in case pod is removed from the host. 
Once a kata pod is assigned to the host, device plugin reserves unused cores for the pod, then modify pod annotations by adding 'io.katacontainers.config.hypervisor.default_vcpus' and 'io.katacontainers.config.hypervisor.cpu_core_list' (a new annotation to be supported by kata).  Upon receiving such annotations, kata shim should start KVM threads per the guidance of 'default_vcpus' and pin these threads to the cores listed in 'cpu_core_list'.

**Describe alternatives you've considered**
How do device plugin notifies kata the core numbers. 
1) Using annotation
2) Add a new API to kata runtime and device plugin makes a call to kata runtime. 
1# is preferred as it decouples the tight integration between device plugin and kata. 

**Additional context**
Pod spec
---
apiVersion: v1
kind: Pod
metadata:
  name: demo-pod
spec:
  containers:
    - name: demo-container-1
      image: k8s.gcr.io/pause:2.0
      runtimeClass: kata-qemu
      resources:
        limits:
          hardware-vendor.example/isolatedcpu: 2

Pod annotations passed to kata
      annotations:
        io.katacontainers.config.hypervisor.default_vcpus: '2'
        io.katacontainers.config.hypervisor.cpu_core_list: '3,4'

**Comment by fengwang666** [comment]
@kata-containers/architecture-committee 
We'd like to get feedback on the approach itself and whether this is something can be upstreamed.

**Comment by liyimeng** [comment]
Love the feature, at least :D

**Comment by liubin** [comment]
In some co-location systems, batch jobs(in Kata Containers) and RPC/HTTP services(in runc) are allocated to different CPUs to reduce performance interference,  I want to give a `+1`.

**Comment by fengwang666** [comment]
@fidencio That's fair. We can talk about this next week. cc @fredliu-db

**Comment by klihub** [comment]
Isn't one problem with this setup that, since you are allocating CPU as a `device` for your workload, you end up not allocating any native CPU resources for them ? Then, even if you do allocate one or more full isolated CPUs for a workload, this never gets assigned to the 'Guaranteed' QoS class (where it really should belong). Therefore it becomes a target for eviction if the node runs low on resources. Regardless of the fact that you want to allocate isolated CPUs for it, which suggests that it is a precious workload which should be protected from eviction.

**Comment by fengwang666** [comment]
@klihub You're absolutely correct. With this proposal, we won't be able to use the 'Guaranteed' QoS to protect the workload. Instead, we probably need to set the Pod Priority to make sure the target workload pod is evicted at last.
